gtsfm/loader/olsson_loader.py

```diff
@@ -76,20 +81,25 @@ def __init__(
 
         # stores camera poses (extrinsics) and intrinsics as 3x4 projection matrices
         # 'P' array will have shape (1,num_imgs), and each element will be a (3,4) matrix
-        data = loadmat(cam_matrices_fpath)
+        data = scipy.io.loadmat(cam_matrices_fpath)
+
+        if len(data["P"][0]) != self._num_imgs:
+            raise RuntimeError("Number of images found on disk not equal to number of ground truth images.")
 
         # M = K [R | t]
         # in GTSAM notation, M = K @ cTw
         M_list = [data["P"][0][i] for i in range(self._num_imgs)]
 
-        # first pose is identity, so K is immediate given
-        self._K = M_list[0][:3, :3]
-        Kinv = np.linalg.inv(self._K)
+        self._K, _ = verification_utils.decompose_camera_projection_matrix(M_list[0])
+
+        self._wTi_list = []
+        # first pose is not necessarily identity (in Door it is, but not in Palace of Fine Arts)
+        for M in M_list:
+            K, wTc = verification_utils.decompose_camera_projection_matrix(M)
+            self._wTi_list.append(wTc)
 
-        # decode camera poses as:
-        #    K^{-1} @ M = cTw
-        iTw_list = [Kinv @ M_list[i] for i in range(self._num_imgs)]
-        self._wTi_list = [Pose3(Rot3(iTw[:3, :3]), iTw[:, 3]).inverse() for iTw in iTw_list]
+        # GT 3d structure (point cloud)
+        self._point_cloud = data["U"].T[:, :3]
```

Comment:
Any API to use this in GTSFM right now?

Reply:
Yeah, I'm using in the `gtsfm/visualization/view_scene_olsson.py` file now to access the GT 3d structure.